### PR TITLE
all: exclude empty outputs in requests commitment

### PR DESCRIPTION
**Pull request by fjl**
Implements changes from these spec PRs:

- https://github.com/ethereum/EIPs/pull/8989
- https://github.com/ethereum/execution-apis/pull/599

### CODE REVIEW

File: beacon/engine/types.go [hunk 1]
```diff
@@ -265,15 +265,7 @@ func ExecutableDataToBlockNoHash(data ExecutableData, versionedHashes []common.H
...
-		// Put back request type byte.
-		typedRequests := make([][]byte, len(requests))
-		for i, reqdata := range requests {
-			typedReqdata := make([]byte, len(reqdata)+1)
-			typedReqdata[0] = byte(i)
-			copy(typedReqdata[1:], reqdata)
-			typedRequests[i] = typedReqdata
-		}
-		h := types.CalcRequestsHash(typedRequests)
+		h := types.CalcRequestsHash(requests)
```
Comment (rjl493456442) [inline beacon/engine/types.go:268]: I am a bit confused here. 

The requests is supplied by consensus client and originally the requests have no prefix type byte.

Do we change the API spec that the requests must have the prefix type byte now?

File: beacon/engine/types.go [hunk 2]
```diff
@@ -343,20 +335,11 @@ func BlockToExecutableData(block *types.Block, fees *big.Int, sidecars []*types.
...
-			plainRequests[i] = reqdata[1:]
-		}
-	}
-
 	return &ExecutionPayloadEnvelope{
 		ExecutionPayload: data,
 		BlockValue:       fees,
 		BlobsBundle:      &bundle,
-		Requests:         plainRequests,
+		Requests:         requests,
```
Comment (rjl493456442) [inline beacon/engine/types.go:342]: Same here. Originally it's requests without prefix type byte, now it's requests with prefix type byte.

File: beacon/engine/types.go (hunk 1 above)
Comment (fjl) [reply beacon/engine/types.go:268]: Yes it's the change I am making to the engine API spec. Link in description.

File: cmd/evm/internal/t8ntool/execution.go [hunk 3]
```diff
@@ -365,22 +365,21 @@ func (pre *Prestate) Apply(vmConfig vm.Config, chainConfig *params.ChainConfig,
...
-		// EIP-6110 deposits
+		requests = [][]byte{}
+		// EIP-6110
 		var allLogs []*types.Log
 		for _, receipt := range receipts {
 			allLogs = append(allLogs, receipt.Logs...)
 		}
-		depositRequests, err := core.ParseDepositLogs(allLogs, chainConfig)
-		if err != nil {
+		if err := ParseDepositLogs(&requests, allLogs, p.config); err != nil {
```
Comment (MariusVanDerWijden) [inline cmd/evm/internal/t8ntool/execution.go:374]:
```suggestion
		if err := core.ParseDepositLogs(&requests, allLogs, chainConfig); err != nil {
```

**Comment by fjl** [comment]
@lightclient PTAL

**Comment by fjl** [comment]
Should be OK to merge now that EIP changes are in.

File: core/state_processor.go [hunk 4]
```diff
@@ -302,26 +299,32 @@ func processRequestsSystemCall(evm *vm.EVM, requestType byte, addr common.Addres
...
 			if err != nil {
-				return nil, fmt.Errorf("unable to parse deposit data: %v", err)
+				return fmt.Errorf("unable to parse deposit data: %v", err)
 			}
 			deposits = append(deposits, request...)
 		}
 	}
-	return deposits, nil
+	if len(deposits) > 1 {
+		*requests = append(*requests, deposits)
```
Comment (MariusVanDerWijden) [inline core/state_processor.go:327]: I don't really understand this, you're adding an empty deposit upfront, shouldn't you remove it here again?

Comment (MariusVanDerWijden) [reply core/state_processor.go:327]: Nvm just figured it out

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

**Pull request merged**